consul/src/main/java/com/linecorp/armeria/client/consul/ConsulEndpointGroup.java

```diff
@@ -0,0 +1,115 @@
+import com.linecorp.armeria.client.endpoint.DynamicEndpointGroup;
+import com.linecorp.armeria.client.endpoint.EndpointGroup;
+import com.linecorp.armeria.client.retry.Backoff;
+import com.linecorp.armeria.internal.consul.ConsulClient;
+
+/**
+ * A Consul-based {@link EndpointGroup} implementation. This {@link EndpointGroup} retrieves the list of
+ * {@link Endpoint}s from a Consul using {@link ConsulClient} and updates it when the status of the
+ * service by polling.
+ */
+public final class ConsulEndpointGroup extends DynamicEndpointGroup {
+
+    /**
+     * Creates a {@code ConsulEndpointGroup} with only service name. This {@code ConsulEndpointGroup} will
+     * retrieves the list of {@link Endpoint}s from a local Consul agent(using default consul service port).
+     */
+    public static ConsulEndpointGroup of(String serviceName) {
+        return builder(serviceName).build();
+    }
+
+    /**
+     * Creates a {@code ConsulEndpointGroupBuilder} to build {@code ConsulEndpointGroup}.
+     */
+    public static ConsulEndpointGroupBuilder builder(String serviceName) {
+        return new ConsulEndpointGroupBuilder(serviceName);
+    }
+
+    private final ConsulClient consulClient;
+    private final String serviceName;
+    private final ScheduledExecutorService executorService;
+    private final Backoff retryBackoff;
+
+    @Nullable
+    private volatile ScheduledFuture<?> scheduledFuture;
+
+    /**
+     * Create a Consul-based {@link EndpointGroup}, endpoints will be retrieved by service name using
+     * {@link ConsulClient}.
+     *
+     * @param consulClient   a consul client
+     * @param serviceName a service name to retrieve
+     * @param executorService a executorService
+     * @param retryBackoff a retry backoff to check
+     */
+    ConsulEndpointGroup(ConsulClient consulClient, String serviceName, ScheduledExecutorService executorService,
+                        Backoff retryBackoff) {
+        this.consulClient = consulClient;
+        this.serviceName = serviceName;
+        this.executorService = executorService;
+        this.retryBackoff = retryBackoff;
+
+        consulClient.healthyEndpoints(serviceName)
+                    .thenAccept(this::setEndpoints);
+        whenReady().join();
+        scheduledFuture = executorService.schedule(this::update,
+                                                   nextDelayMillis(),
+                                                   TimeUnit.MILLISECONDS);
+    }
+
+    private void update() {
+        if (isClosing()) {
+            return;
+        }
+        consulClient.healthyEndpoints(serviceName).thenAccept(this::setEndpoints);
+        scheduledFuture = executorService.schedule(this::update,
+                                                   nextDelayMillis(),
+                                                   TimeUnit.MILLISECONDS);
+    }
+
+    private long nextDelayMillis() {
+        final long delayMillis = retryBackoff.nextDelayMillis(1);
```

Comment:
If a consul endpoint group gets endpoint lists periodically, we don't have to use backoff.
How about just taking `retryIntervalMillis`?

Reply:
Actually, it is not `retry`. I changed that to `intervalMillis`.

Reply:
Ok, I will change that to `intervalMillis`. Actually, that is not `retry`.